lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+        self:emit_signal("widget::layout_changed")
+    end
+    return ret
+end
+
+
+--- Remove one or more widgets from the layout
+-- @param widget ... Widgets that should be removed (must at least be one)
+-- @treturn boolean index If the operation is successful
+function grid:remove_widgets(...)
+    local args = { ... }
+    local ret = false
+    for _, rem_widget in ipairs(args) do
+        -- loop over the existing widgets
+        for index, data in ipairs(self._private.widgets) do
+            if data.index ~= index then data.index = index end
+            if data.widget == rem_widget then
+                matrix_remove(self._private.matrix, data.row, data.col, data.row_span, data.col_span)
+                table.remove(self._private.widgets, index)
+                ret = true
+                break
+            end
+        end
+    end
+
+    if ret then
+        -- Recalculate num_rows and num_cols
+        self:get_shape()
+        self:emit_signal("widget::layout_changed")
+    end
+    return #args > 1 and ret
+end
+
+--- Return the widget at coordinates
+-- @param number row The row coordinate of the widget
+-- @param number col The column coordinate of the widget
+-- @return widget The widget found at the specific coordinates, otherwise nil
+function grid:get_widget_at(row, col)
+    if not row or row < 1 or row > self._private.num_rows then return nil end
+    if not col or col < 1 or col > self._private.num_cols then return nil end
+    for _, data in ipairs(self._private.widgets) do
+        if row > data.row - 1 and row < data.row + data.row_span and
+           col > data.col - 1 and col < data.col + data.col_span then
+            return data.widget
+        end
+    end
+    return nil
+end
+
+-- Return list of children
+function grid:get_children()
+    local ret = {}
+    for _, data in ipairs(self._private.widgets) do
+        table.insert(ret, data.widget)
+    end
+    return ret
+end
+
+-- Add list of children to the layout
+function grid:set_children(children)
+    self:reset()
+    if #children > 0 then
+        self:add(unpack(children))
+    end
+end
+
+--- Set the preferred orientation of the grid layout
+--@DOC_wibox_layout_grid_orientation_EXAMPLE@
+-- @property string orientation
+function grid:set_orientation(val)
+    if self._private.orientation ~= val and (val == "x" or val == "y") then
+        self._private.orientation = val
+    end
+end
+
+--- Force the number of columns of the layout
+-- @property forced_num_cols
+function grid:set_forced_num_cols(val)
+    if self._private.forced_num_cols ~= val then
+        self._private.forced_num_cols = val
+        self._private.num_cols = val
+        self:emit_signal("widget::layout_changed")
+    end
+end
+
+--- Force the number of rows of the layout
+-- @property forced_num_rows
+function grid:set_forced_num_rows(val)
+    if self._private.forced_num_rows ~= val then
+        self._private.forced_num_rows = val
+        self._private.num_rows = val
```

Comment:
This is wrong. It should call `self:get_shape()` instead, because the grid could already have more rows than the new argument dictates. Otherwise, the next call to `get_shape` will suddenly increase the number of rows (without emitting any signals)